### ISSUE TITLE
Strange bug: cdist segfaults on M1 Max ARM, for certain size tensors, if numpy is loaded first

### ISSUE THREAD

**Issue by battaglia01** [pytorch/pytorch#130230]
### 🐛 Describe the bug

There seems to be an issue where `cdist` will segfault on M1 Max for certain input sizes only, and only if numpy is loaded first. For instance, this is sufficient to segfault on my system

```
import numpy as np
import torch

t = torch.randn(18, 7)
dist_matrix = torch.cdist(t, t) # <-- Segmentation fault here
```

But somehow, this doesn't, even though it is a larger tensor

```
import numpy as np
import torch

t = torch.randn(30, 7)
dist_matrix = torch.cdist(t, t) # <-- No segmentation fault
```

If I import torch before numpy, this problem goes away

```
import torch
import numpy as np

t = torch.randn(18, 7)
dist_matrix = torch.cdist(t, t) # <-- No segmentation fault
```

And if I set the number of threads lower, the chances of segfaulting are less. For 1 thread it never segfaults, for 10 it always does, and for 5 it seems to have a 50/50 chance:

```
import numpy as np
import torch

torch.set_num_threads(1)
t = torch.randn(18, 7)
dist_matrix = torch.cdist(t, t) # <-- No segmentation fault, usually
```

This is with numpy 1.26.4, which was the very last version released before 2.0.0. 2.0.0 was released about a month before this writing and I haven't upgraded to it yet. I'm not sure if anything would change with that.

### Versions

Collecting environment information...

PyTorch version: 2.3.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (http://root.cern/git/clang.git 25f804797f80b69d8c56794e3e0300acd9458958) (http://root.cern/git/llvm.git ac712f0f44b45a1455a302dd6cbb7b6cce269d2d)
CMake version: version 3.29.5
Libc version: N/A

Python version: 3.11.9 | packaged by conda-forge | (main, Apr 19 2024, 18:34:54) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] onnxruntime==1.18.0
[pip3] torch==2.3.1
[conda] libopenvino-pytorch-frontend 2024.1.0             h00cdb27_7    conda-forge
[conda] nomkl                     1.0                  h5ca1d4c_0    conda-forge
[conda] numpy                     1.26.4          py311h7125741_0    conda-forge
[conda] torch                     2.3.1                    pypi_0    pypi

**Comment by battaglia01** [comment]
After going to 2.2.2 this doesn't happen anymore.

**Comment by albanD** [comment]
Sounds like an OMP-style issue due to numpy symbols.
Unfortunately, numpy sometimes loads shared libraries in a global way which then prevent pytorch from loading it's own "clean" version.

As a general rule, I would suggest to always import torch first to avoid this kind of issues.
cc @malfet was there a change in OMP packaging for 2.3?

**Comment by malfet** [comment]
I really hope that this has been fixed in trunk/upcoming 2.4

**Comment by albanD** [comment]
@battaglia01 would you be able to install the 2.4 rc as described in https://dev-discuss.pytorch.org/t/pytorch-release-2-4-0-final-rc-is-available/2213 and test if the same still happens?

**Comment by battaglia01** [comment]
Ran it, no problem with the new 2.4 rc!

**Comment by malfet** [comment]
>  was there a change in OMP packaging for 2.3?

Yes, for a while PyTorch for Apple Silicon has been built without OpenMP support and starting with 2.3 it was enabled

**Comment by albanD** [comment]
This is now fixed in main and will be in the next release. So closing the issue.
